Add documentation for Adreno deployment

dsbarinov1 opened the pull request:
@echuraev @elvin-n review please :)

dsbarinov1 committed:
Add documentation for Adreno deployment

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+Build and deploy model for Adreno
+---------------------------------
+
+As the deployment pipelines on Android and on Adreno do not have many
```
Adreno is the part of the Android, so it is obviously that they "do not have many differences", because the process is the same.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+aspect to compile and deploy models for Adreno and will take a look at
+the generation of kernels with and without textures. In addition, the
+possibility of choosing a different precision for model compilation will
+be considered.
```
Probably we can also add a link to the how to guide in the beginning of this section.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+---------------------------------
+
+As the deployment pipelines on Android and on Adreno do not have many
+differences, in this section we will focus on one and only necessary
+aspect to compile and deploy models for Adreno and will take a look at
```
"on one and only necessary aspect to compile and deploy models for Adreno" sounds a bit difficult for understanding after the first reading. What is this aspect?

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+be considered.
+
+| |Android deployment pipeline|
+| Fig.2 Deployment pipeline on Adreno devices ### Adreno target
+  Normally, when compiling models on Android using OpenCL, the
+  corresponding target is used
```
Looks like there are some problems with formatting.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+
+   target="opencl -device=adreno"
+
+Let’s write simple model and take a look at generated kernels for these
+two targets
```
Probably add a couple of words about this simple model. What operation is used in this model?

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+possibility of choosing a different precision for model compilation will
+be considered.
+
+| |Android deployment pipeline|
```
Align all down arrows to have the same sizes here:
<img width="336" alt="image" src="https://user-images.githubusercontent.com/5525113/197483405-71363f04-37f1-428a-bf0f-df0f7a851130.png">

Be sure that all arrows are direct and don't have angles like here: 
<img width="640" alt="image" src="https://user-images.githubusercontent.com/5525113/197483616-b6bd309b-00a4-4210-ad62-59f90e755348.png">

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+   "weight": tvm.nd.array(filter)
+   }
+
+Classical opencl target:
```
Probably also add a bit more details about the next part of the code

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+
+.. code:: python
+
+   target="opencl"
+
+   with tvm.transform.PassContext(opt_level=3):
+   graph, lib, params = relay.build_module.build(mod, target, params=params)
+   print(lib.imported_modules[0].get_source())
```
Be careful about the indents. This code won't work in the way as you expect.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+   graph, lib, params = relay.build_module.build(mod, target, params=params)
+   print(lib.imported_modules[0].get_source())
+
+Notice the generated convolution kernel and the presence of pointers in
+the initialization of the function:
```
`the presence of pointers` - What does it mean? I see the OpenCL code first time and I don't understand why it is important.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+   target="opencl -device=adreno"
+
+   with tvm.transform.PassContext(opt_level=3):
+   graph, lib, params = relay.build_module.build(mod, target, params=params)
+   print(lib.imported_modules[0].get_source())
```
indents

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+   graph, lib, params = relay.build_module.build(mod, target, params=params)
+   print(lib.imported_modules[0].get_source())
+
+We can now observe the use of textures in the initialization of the
+function:
```
Initialization, is it the only one difference between buffers and textures?

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+~~~~~~~~~~
+
+We can also set different precision, choosing from *float16*,
+*float16_acc32* (Mixed Precision), *float32*. First of all, we need to
+register conversion to mixed precision
```
Add more details. Imagine, I'm a new user and I don't understand this code.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+           mixed_precision_type,
+       ]
+
+We then need to obtain **mod**, after which we can convert it to
```
What is the `mod`?

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+           seq = tvm.transform.Sequential(
+               [
+                   relay.transform.InferType(),
+                   relay.transform.ToMixedPrecision() # primary method
```
What is the `ToMixedPrecision`? Add a couple of words about it.

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,259 @@
+The complete step-py-step process of compiling and deploying models on
+Adreno, including selection of precision, running the inference of the
+model, getting the predictions, and measuring the performance can be
+found in this tutorial: `How To <>`__
```
Why you didn't write the correct link address?

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+#   pip install mxnet
+#
+# Besides that, you should have TVM builded for Android.
+# Please, refer to: `Compile and deploy models on Adreno GPU <link>`_ .
```
Where is the link?

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+# Then to upload these two files to the device you should use:
+# .. code-block:: bash
+#
+#   adb -s <device_hash> push {libtvm_runtime.so,tvm_rpc} /data/local/tmp_
```
```suggestion
#   adb -s <device_hash> push {libtvm_runtime.so,tvm_rpc} /data/local/tmp
```

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+#
+# At this moment you will have «libtvm_runtime.so» and «tvm_rpc» on path /data/local/tmp on your device.
+# Sometimes cmake can’t find «libc++_shared.so». Use:
+# .. code-block:: bash
+#
+#   find ~/Android/sdk/ndk/ -name libc++_shared.so
+#
+# (Linux)
+# .. code-block:: bash
+#
+#   find ~/Library/Android/sdk/ndk/ -name libc++_shared.so
+#
+# (MacOS)
```
```suggestion
# .. code-block:: bash
#
#   find ${ANDROID_NDK_HOME} -name libc++_shared.so
#
```

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+# Load pretrained MXNet model
+# ---------------------------
+# Create a Relay graph from a MXNet VGG16 model
+import  os
+import  numpy  as  np
+import  mxnet.gluon  as  gluon
+import  tvm
+from  tvm  import  relay, rpc
+from  tvm.contrib  import  utils, ndk
+from  tvm.contrib  import  graph_executor
```
```suggestion
import os
import numpy as np
import mxnet.gluon as gluon
import tvm
from tvm import relay, rpc
from tvm.contrib import utils, ndk
from tvm.contrib import graph_executor
```

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+# Specify Adreno target before compiling to generate texture 
+# leveraging kernels and get all the benefits of textures 
+
+target="opencl -device=adreno"
+target_host="llvm -mtriple=arm64-linux-android"
```
Should we add a description about `target_host`?

elvin-n replied:
BTW, this is depricated usage of the target/target host. Currently you have to create python object Target and initialize it by apprpriate target_host

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+######################################################################
+# Load a test image
+# -----------------
+# As an example we would use classical cat image from ImageNet
```
Display in the document this image.

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,298 @@
+#    Evaluate inference time cost...
+#    Execution time summary:
+#     mean (ms)   median (ms)    max (ms)     min (ms)     std (ms)  
+#      76.6101      76.6053      77.6530      75.5720       0.6789
```
Did you try to convert this script into the web page?

elvin-n replied:
It must not be converted manually. It is done automatically when documentation is built

echuraev replied:
Yes, I know, but probably it would be better to convert it locally and be sure that everything is fine in the final document.

elvin-n replied:
the main value is py script that is potentially can be executed by user. Why do we care about doc? It should be converted automatically...

echuraev replied:
Because from this .py script, web page will be automatically generated and presented in the official TVM documentation, and I want to be sure that everything looks correctly. This is why I asked this question. I don't think that there is the problem to build this page locally and just check that everything is fine.

dsbarinov1 committed:
Fix text, fix links, add images

dsbarinov1 committed:
Fix text, fix links, add images

dsbarinov1 commented:
@echuraev @elvin-n some fixes done

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,266 @@
+Precisions
+~~~~~~~~~~
+
+We can also set different precision, choosing from *float16*,
```
->
To leverage the GPU hardware capabilities and utilize benefits of half precision computation and memory management, we can convert original model having floating points operation to model operating with half precision. Choosing lower precision will positively affect the performance of the model, but it may also have a decrease in the accuracy of the model.

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,266 @@
+In some tasks we may sacrifice accuracy in favor of speed, and in some tasks we may prefer mixed precision,
+obtaining some kind of balance.
+
+First of all, to be able to convert precisions in general, we need to
+register conversion to mixed precision
+
+.. code:: python
+
+   from  tvm.relay.op  import  register_mixed_precision_conversion
+
+   conv2d_acc = "float32"
+
+   # Pick a priority > 10 to overwrite defaults, higher priorities take precedence
+   @register_mixed_precision_conversion("nn.conv2d", level=11)
+   def  conv2d_mixed_precision_rule(call_node: "relay.Call", mixed_precision_type: str):
+       global  conv2d_acc
+       return [
+           # always do main calculation in mixed_precision_type
+           relay.transform.mixed_precision.MIXED_PRECISION_ALWAYS,
+           # the dtype for the accumulator
+           conv2d_acc,
+           # the output dtype for the operation (usually fp16)
+           mixed_precision_type,
+       ]
+
+   # Same for dense
+   @register_mixed_precision_conversion("nn.dense", level=11)
+   def  conv2d_mixed_precision_rule(call_node: "relay.Call", mixed_precision_type: str):
+       global  conv2d_acc
+       return [
+           relay.transform.mixed_precision.MIXED_PRECISION_ALWAYS,
+           conv2d_acc,
+           mixed_precision_type,
+       ]
+
```
it is not fully true. these functions are required *only* for mixed precision or for generalization of the pure fp16/mixed precision. For straightforward conversion into pure fp16 they are not required

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,266 @@
+               mod = seq(mod)
+       return  mod
+
+   dtype="float16_acc32"
```
I would not like mention mixed precision in the first presentatoin of ToMixedPrecision at all. If we want to cover mixed precisoin, let's describe conversion into fp16 and then have a section with regestring functions, etc, etc

elvin-n commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,281 @@
+
+.. code-block:: bash
+
+  pip install mxnet
```
this is quite bad choice of the model. It has benefits as ease-of-use, but tvm has issue with latest python and latest mxnet: https://github.com/apache/tvm/issues/11452

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,266 @@
+*float16_acc32* (Mixed Precision), *float32*. Choosing lower precision may positively
+affect the performance of the model, but it may also have a decrease in the accuracy of the model.
+In some tasks we may sacrifice accuracy in favor of speed, and in some tasks we may prefer mixed precision,
+obtaining some kind of balance.
```
Need to add what mixed precision mean. May be something like this

```
ToMixedPrecision pass traverse over the network and split network to clusters of ops dealing with float or float16 data types. The clusters are defined by three types of operations
* Operations always be converted into float16 data type
* Operations which can be converted if they follow by converted cluster
* Operations never be converted to the float16 data type  
This list is defined in the ToMixedPrecision implementation https://github.com/apache/tvm/blob/main/python/tvm/relay/transform/mixed_precision.py#L34 and can be overridden by user

In some cases we want to have accumulation type in data type with bigger type of bits that input data type. This is supported, for example, for conv2d and dense operations. To override accumulation type you need to register function with @register_mixed_precision_conversion decorator to modify parameters of ToMixedPrecision conversion
```

dsbarinov1 committed:
Fix docs

dsbarinov1 commented:
Please review "Fix docs" commit @elvin-n @echuraev

Commit:
[COMMUNITY] Jyotsna Verma -> Reviewer (#13251)

adding Jyotsna to reviewers list

driazati committed:
[skip ci] Revert "fix GPU other build (#13235)" (#13261)

This reverts commit e9e8c4b263aa33bfd6bd54151a4bc7eb89da8deb.

masahi committed:
[MetaSchedule] Swap the order of RewriteTensorize and VerifyGPUCode to reduce tuning time (#13259)

* [MetaSchedule] Swap the order of RewriteTensorize and VerifyGPUCode to
reduce tuning time

* add comment

lhutton1 committed:
[CI] Skip failing Caffe tests due to broken URL (#13228)

See issue #13227.

Co-authored-by: driazati <9407960+driazati@users.noreply.github.com>

echuraev commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,315 @@
+
+   __kernel void tvmgen_default_fused_nn_conv2d_kernel0(__write_only image2d_t pad_temp_global_texture, __read_only image2d_t p0) {
+   // body..
+
```
I would add some more details about the differences, e.g.:

`image2d_t` is a built-in OpenCL types that represents two-dimensional image object and provides several additional functions. Also, when we use `image2d_t` we read 4 elements at one time, and it helps to utilize hardware in a more efficient way.

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,295 @@
+"""
+
+
+#################################################################
```
Why before you use `"""` for commenting and now `#`? Probably use one comment style for huge multiline comments?

echuraev commented on gallery/how_to/deploy_models/deploy_model_on_adreno.py:
```diff
@@ -0,0 +1,295 @@
+#
+# .. code-block:: bash
+#
+#   export TVM_TRACKER_HOST=0.0.0.0
+#   export TVM_TRACKER_PORT=9190
+#
+# check that the tracker is running and the device is available
+#
+# .. code-block:: bash
+#
+#     python -m tvm.exec.query_rpc_tracker --port 9190
```
IMHO, using environment variables is not the most convenient way to configure tracker. You can also pass all these configurations through command line.

lhutton1 committed:
[TVMC] Apply constant folding when converting layout (#13216)

This commit ensures that constant folding is applied when a desired
layout is selected during compilation. It ensures that
`layout_transform` operations are removed where possible so that
pattern matching for BYOC backends can work effectively.

A test has been added to check this regression.

lhutton1 committed:
Apply group write permissions to Python virtual environment (#13252)

This commit applies additional write permission to the "tvm-venv"
group virtual environment. Currently after entering a container from
a newly built image it dosn't seem possible to install/update Python
packages. E.g. updating pip will give errors such as:
```
$ pip install --upgrade pip
ERROR: Could not install packages due to an OSError: [Errno 13]
Permission denied: '/venv/apache-tvm-py3.7/bin/pip' Check the
permissions.
```

Enabling write access for this group fixes this as long as the
current user is a member of the "tvm-venv" group.

yelite committed:
Fix a typo in rpc/client.py (#12842)

mehrdadh committed:
[Hexagon] Add pylint on tests (#13233)

* [Hexagon] Tests pylint

* fix error

* Fix buffer name

Commit:
[build][relay][te][tir] remove unused vars / args (#13266)

- Fix clang 15.0.3 '-Wunused-but-set-variable' and '-Wunused-lambda-capture' warnings by removing / commenting-out code.

apivovarov committed:
[Frontend][Tensorflow2] Import graph_def to default graph before calling function_def_to_graph_def (#13260)

[TF2] Import graph_def to default graph before calling function_def_to_graph_def

woobinw committed:
[Frontend][PaddlePaddle] Fix UnboundLocalError: local variable 'shape… (#13247)

There are a local variable referenced before assignment in convert_interpolate function. I think varible 'size' is real want to be referenced.

driazati committed:
[skip ci] Revert "[ci] Protect release branches (#13208)" (#13274)

This reverts commit 5acf3f90c63b6760cd23796b442f8ac20e645af0.

Reverting since this is causing some spam from the ASF Infra bot related
to https://issues.apache.org/jira/browse/INFRA-23834. As in that issue
the protections have been applied manually by ASF Infra so this revert
shouldn't have any real effect

dsbarinov1 committed:
Fix docs v2

dsbarinov1 commented:
@echuraev small fixes done

bkmgit committed:
[Docs] Minimal dependencies for Fedora/CentOS (#13248)

Minimal dependencies for Fedora/CentOS

This commit indicates how to install minimal set of
dependencies for building Apache TVM on Fedora and
CentOS. It supplements existing information for
Ubuntu and MacOS.

Commit:
[build][doc] Fix clang doxygen warnings (#13270)

Fix occurrences of clang's `-Wdocumentation-unknown-command` warning.

Commit:
[build][tir] fix clang redundant-move warning (#13268)

Fix code to address a valid `-Wredundant-move` clang warning.

lhutton1 committed:
[ETHOSN] Inline non-compute-intensive partitions (#13092)

* [ETHOSN] Inline non-compute-intensive partitions

Adds a pass that analyzes functions partitioned for the NPU and inlines
those that are deemed "non-compute-intensive" back to the main function
so that they can be considered for other backends. The current heurisic
for deciding a non-compute-intensive function is to collectively check
all of the operations in the function have no multiply accumulate
operations. This heuristic is not optimial; optimization is left for
future exploration.

This pass is inspired by the "IsComputeIntensiveGraph" pass in the
TensorRT integration.

Change-Id: I20c197702f5252f102cfc1e4b4635ab836aa7835

* Address comments

* 'inline_non_compute_intensive_partitions' -> 'is_inline_non_compute
_intensive_partitions_enabled'.
* remove no MAC operations.
* fix network test.

Change-Id: Ie1015b27f37e47544bed6f0aff819ee4649de579

* Fix failing unit tests due to optimization

Change-Id: I0ee0af071dc77c91e0ef0f6753506cb40d1d1859

* Add future exploration suggestions

Change-Id: Ie918d7f1059f032282f1f5eeffda38f4febcd59c

lhutton1 committed:
[ETHOSN] Throw error message when inference fails (#13022)

* [ETHOSN] Throw error message when inference fails

Previously the runtime would silently skip interence failures and return
random values as the result. This can make spotting inference failures
challenging. The runtime now throws a fatal error when inference did not
complete successfully along with an error message that gives some
details about the error that occurred.

Change-Id: Iadb6da04ad1c906e3ec49959eb3da0978295aebf

* Address comments

* clarify test file brief
* add test case for running status
* add driver stack reference to WaitStatus class

Change-Id: I792742892b761534904816135ae2ffcb3f028b2c

zxybazh committed:
[MetaSchedule] Fix Task Hanging in EvolutionarySearch (#13246)

This PR introduces a new argument for EvolutionarySearch that limits the failures (defined as rounds of no new generated candidate) in the `SampleInitPopulation` stage. In this way we can avoid the task to be hanging forever in special cases, e.g., some postproc always fails. This should fix #12330.

sunggg committed:
[Bugfix][TIR] Fix version conflict with `typing` for Python 3.9  (#13269)

Current type checker for TIR schedule had issue with typing for Python 3.9. 
This simple patch fixes this problem.

yelite committed:
[MetaSchedule] Improve the script for TorchBench model tuning & benchmarking (#13255)

This PR adds features to the `python/tvm/meta_schedule/testing/torchbench/run.py`.

- Integrate with the TVM PyTorch integration to handle boolean tensor and unaligned memory.
- Deduplicate collected tuning tasks to prevent thousands of tasks created by hundreds of subgraphs with similar structure.
- Add option to cast model to float32, which are more stable numerically than float16 and prevents inaccurate result from many models.
- Add option to choose search strategy in MetaSchedule.
- Inspect output error if the actual output doesn't match the expectation. Also save the actual output and expected output for further analysis if needed.
- Save subgraphs and their example input for debug purpose.
- Print MetaSchedule profiling information at the end of execution.
- Detach PyTorch tensor before exporting to dlpack.
- Fix the sys path to avoid conflict with the `benchmarks` package installed by TorchBench dependency.
- Trim all command line args passed in, in order to prevent breaking some TorchBench model that depends on args.
- Empty cuda cache before starting the actual benchmark.

wzh99 committed:
[Relay] Add tensor rank check for `nn.instance_norm` (#13280)

Add tensor rank check for `nn.instance_norm`.

leiwen83 committed:
[Relay] Enhancement for fold_scale_axis and simplify_expr (#13275)

add(%1, %1) convert to multiply(%1, 2f); enhance fold_scale_axis to fold multiply(%1, 2f) into conv

Signed-off-by: Lei Wen <wenlei03@qiyi.com>
Co-authored-by: Lei Wen <wenlei03@qiyi.com>

tqchen committed:
[skip-ci][COMMUNITY] New committer Ashutosh Parkhi (#13286)

[COMMUNITY] New committer Ashutosh Parkhi

Lunderberg committed:
[TIR][Arith] Use TryCompare to narrow inequalities if possible (#13024)

Prior to this commit, the result of TryCompare would only be used if
it could definitively prove a conditional to be either true or false.
For example, if it is known that `0 <= i`, a conditional of `i <= 0`
would be left as-is.

This commit introduces rewrite rules to preferentially simplify
into more restrictive conditions.  Using the same example, if it is
known that `0 <= i`, a conditional of `i <= 0` would be simplified
into `i == 0`.  Similarly, if it is known that `0 <= i`, a
conditional of `i != 0` would be simplified into `0 < i`.

Because this change does not introduce significant overhead, as the
results of `RewriteSimplifier::Impl::TryCompare` are already
available, this change is enabled for all use cases and does not
require a call to `RewriteSimplifier::SetEnabledExtensions`.

Commit:
[build][hexagon] remove unused variable (#13291)

Remove unused member variable in the `SimulatorRPCChannel` class.
Fixes a clang warning.

liangW-intellif committed:
[BugFix][Pattern] Fixed a crash when AltPattern and FunctionPattern are used nested (#13278)

The PatternGroup doesn not check if the FunctionPattern is matched
while processing the FunctionPattern, but when FunctionPattern
is nested with AltPattern, the FunctionPattern may not be matched,
resulting in a crash when looking up matched nodes.
This commit adds a check at handling FunctionPattern to fix this crash.

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+Advantages of the Textures
+--------------------------
+
+One of the Adreno's advantages is clever handling of textures. At
```
@dsbarinov1  *the clever handling of textures

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+   fetched from level-2 (L2) cache for texture operations (primary
+   reason)
+
+-  Handling of image boundaries is built-in.
```
*The handling of

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+-  Supports numerous image format and data type combinations with
+   support for automatic format conversions
+
+In addition, Adreno GPUs support up to 32 total textures in a
```
@dsbarinov1 Do we need to mention this paragraph? I think we should delete it. 

@echuraev Does OpenCL expose this 32/16 texture #s? Does it change how we develop schedules? 

If not, we don't need to specify these 2 paragraphs.

echuraev replied:
Actually, we don't use these features, which are described in these two paragraphs. I also had doubts if we really need to mention this in the documentation.

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+
+where **N** is the number of cores available on your *CPU*.
+
+At this stage you have builded TVM for Adreno.
```
*you have built

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+Adreno target
+~~~~~~~~~~~~~
+
+Normally, when compiling models on Android using OpenCL, the
```
for Android

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+
+   target="opencl -device=adreno"
+
+Let's write simple model with one convolutional (conv2d) layer and take a look at generated kernels for these
```
*write a simple model

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+
+Precisions
+~~~~~~~~~~
+The right choice of precision for a specific task can greatly increase the efficiency of the solution,
```
Replace task -> model or workload.

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+
+**Float16**
+
+To leverage the GPU hardware capabilities and utilize benefits of half precision computation and memory management,
```
* the benefits of half

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+**Float16**
+
+To leverage the GPU hardware capabilities and utilize benefits of half precision computation and memory management,
+we can convert original model having floating points operation to model operating with half precision.
```
* an original model 
* to a model operating

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+`relay/transform/mixed_precision.py <https://github.com/apache/tvm/blob/main/python/tvm/relay/transform/mixed_precision.py#L34>`_ 
+and can be overridden by user
+
+In some cases we want to have accumulation type in data type with bigger type of bits that input data type.
```
Reword:
In some cases, we want higher precision in accumulation than the input data

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,319 @@
+FP16 or FP32 accumulation dtypes). Doing this transformation is useful for reducing model size
+as it halves the expected size of the weights (FP16_acc16 case).
+
+From this point we can compile our model as normal
```
*From this point onwards,

TejashShah commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,315 @@
+           mod, target_host=target_host, target=target, params=params
+       )
+
+.. |High-level overview of the Adreno A5x architecture for OpenCL| image:: images/adreno_architecture.png
```
+1

TejashShah reviewed (changes requested):
@dsbarinov1 Thanks for putting together this doc. Please address the comments from Egor and me. 

I liked the fact that you covered build TVM and compile models instructions. As well, how to run any model in different precision. Plus, a sample script for resnet-18 model that could be executed by reader e2e.

Commit:
[build][tir] suppress -Woverloaded-virtual warning (#13267)

- Address a (valid) warning from  clang-15.0.3 regarding the
  `tvm::tir::DataTypeRewriter` class.

- Make some class methods `protected` rather than `public`
  to better reflect authors' intent.

nverke committed:
[Tensorize] Add logs to comparator to make debugging tensorize failures easier (#13285)

* [TIR][Tensorize] Add error logs to IR comparator to display what caused tensorization to fail

* lint issues

mehrdadh committed:
[Hexagon] Lint tests part 2 (#13271)

* Hexagon test lint part 2

* fix import

* fix global variable

* fix import issue

* fix import

* fix exception error

* address comments

masahi committed:
[TE] Make `elem_offset` of the buffers created by `te.extern` a variable to avoid crash (#13297)

* make elem_offset of the buffers created by te.extern a variable

Co-authored-by: Eric Lunderberg <elunderberg@octoml.ai>

* add test

* fix te extern create_prim_func test

Co-authored-by: Eric Lunderberg <elunderberg@octoml.ai>

Lunderberg committed:
[TIR] Added unit test for dynamic parameter in layout transform (#13298)

wrongtest-intellif committed:
[TIR] Preserve loop annotation after loop partitioning (#13292)

Preserve loop annotations when the loop is get partitioned. Also we bind the loop region info to the analyzer for some cases some partition condition could not get solved due to unknown (but trivial) loop region.

Commit:
[FIX] Handle matmul where one inner dimension is unknown (#13287)

Unify the two inner dimensions in the type checker so if one is unknown
it will be filled in.

StrongerXi committed:
[DOCS][TVMC] Use correct argument to reuse tuning records (#13302)

Update tvmc tutorial code to use correct argument for reusing tuning    
records. Specifically, current code uses tuning_records, which is meant 
for saving the generated tuning results, not reusing prior results. We  
should use prior_records instead.

quic-sanirudh committed:
[Hexagon] Fix Hexagon external libs check (#13257)

When building tvm runtime with hexagon we face the below error if
USE_HEXAGON_EXTERNAL_LIBS is not defined. This happens because
USE_HEXAGON_EXTERNAL_LIBS=OFF is defined as the default in
CMakeLists.txt. The modified condition can check for all cases including
undefined variable, empty string and OFF

CMake Error at cmake/modules/Hexagon.cmake:203 (message):
  Invalid use of USE_HEXAGON_EXTERNAL_LIBS=OFF; USE_HEXAGON_EXTERNAL_LIBS
  only supports absolute paths and git repository urls
Call Stack (most recent call first):
  CMakeLists.txt:477 (include)

Commit:
[Relay][Op] Add support for large index fp16 mean and var (#13289)

Add support for large index fp16 mean and var.

Wanger-SJTU committed:
[Bugfix][Runtime] Fix sched_setaffinity in Android (#13158)

* fix sched_setaffinity error on Android

* fix sched_setaffinity error on Android

* fix sched_setaffinity error on Android

* clang format

* add ndk api verion macro

* clang format

masahi committed:
[Torch] Fix advanced indexing with boolean mask (#13306)

* [Torch] Fix advanced indexing with boolean mask

* add comment

mengceng15 committed:
[TOPI] Enhance VNNI dot product (#12588)

change

junrushao committed:
[MetaSchedule] Refactor ScheduleRule Attributes (#13195)

woobinw committed:
[Frontend][PaddlePaddle] Add test case for interpolate op convert func… (#13277)

Add test case for interpolate op convert function #13247

StrongerXi committed:
[BugFix][Driver] Correctly propogate simple-mode flag in LowerSchedule (#13311)

Currently one version of `tvm::LowerSchedule` doesn't pass along the input `simple_mode` flag, which causes it to default back to `false`. This commit fixes it by passing along the input flag.

mehrdadh committed:
[microTVM] Fix RPC session close on runtime side (#13310)

Currently, the RPC session on C/C++ side does not know if the session  
was closed on Python side which causes extra read/write on transport   
while the session is already closed. This commit reuses the Hexagon    
approach in microTVM to shutdown the RPC session.

janetsc committed:
[Hexagon] [runtime] Move lock/unlock to HexagonHtp temporarily (#13318)

Move lock/unlock to HexagonHtp temporarily

yelite committed:
[TIR] Add thread sync if access index doesn't depend on thread index (#13314)

This PR updates the `src/tir/transforms/thread_storage_sync.cc`, to make it insert storage sync if the access index doesn't depend on the innermost thread index, i.e., being constant wit respect to the innermost thread id. 

This fixes an accuracy problem on model https://github.com/pytorch/benchmark/tree/main/torchbenchmark/models/timm_efficientdet

lhutton1 committed:
[ETHOSN] Consolidate target string usage (#13159)

* [ETHOSN] Consolidate target string usage

Removes support for a deprecated target string. The deprecation warning
has been around for a couple of releases now so it should be safe to
remove. The target to use moving forward is: `ethos-n -variant=n78 ...`

Refactored direct use of a driver stack target string in the testing
infrastructure to use the same string we expect users to provide. This
simplified some of the code in codegen and hopefully avoids confusion
in the future.

echuraev committed:
[Adreno][Textures] Fix static memory planner  (#13253)

* [Adreno][Textures] Fix static memory planner

Fix memory reusage in static memory planner.

* Move token allocators to separate file

* Add test on TokenAllocator2d

* Apply comments and fix CI

dsbarinov1 committed:
Release

echuraev reviewed (approved):
LGTM. Thanks

leiwen83 committed:
Fixup libtorch backend build (#13320)

Add clang-format disable for header to prevent reorder.
Torch header file need to be put at the end since torch's dlpack
is a little different with tvm's.

Signed-off-by: Lei Wen <wenlei03@qiyi.com>
Co-authored-by: Lei Wen <wenlei03@qiyi.com>

Lunderberg committed:
[TVMScript] Hide trailing return type if None (#13308)

Because the majority of TIR PrimFuncs operate on buffers, write
their outputs to an output parameter, and do not return a value,
the `-> None` in the function signature becomes visual noise.
This commit removes printing of the return type in cases where
the PrimFunc has no return value.

echuraev committed:
[OpenCL][unit tests] Fix opencl cpp unit tests (#13254)

* [OpenCL][unit tests] Fix opencl cpp unit tests

After some changes in Hexagon, the run of cpp opencl tests leads to the
following error:
```
pluggy.manager.PluginValidationError: unknown hook 'pytest_configure_node' in plugin <module 'tvm.contrib.hexagon.pytest_plugin'
```
Added `pytest_plugin` for OpenCL CPP tests for avoiding this error and
processing gtest arguments.

* Fix fail than gtest_args option was already added

* Move `gtest_args` deginition to the main testing plugin

mehrdadh committed:
[microTVM][CRT] Add memory size as project option (#13313)

* Add memory size as project option

* cleanup

* address comments

* address comments

vinx13 committed:
[TIR] Remove redundant add in vnni/arm intrin (#13319)

* [TIR] Remove redundant add in vnni intrin

* Update arm intrin

Co-authored-by: Ubuntu <ubuntu@ubuntu.com>

vinx13 committed:
[TIR] Allow folding cast with broadcast and ramp (#13317)

elvin-n committed:
[Doc] Add desc of keep_orig_output_dtype for ToMixedPrecision (#13321)

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,311 @@
+                   relay.transform.ToMixedPrecision()
+               ]
+           )
+           with  tvm.transform.PassContext(opt_level=3):
```
One more note that is not much relevant to this network, but generically it is correct. Need to point relay.ToMixedPrecision.keep_orig_output_dtype explicitly during conversion:

```
        with tvm.transform.PassContext(
                config={"relay.ToMixedPrecision.keep_orig_output_dtype": True},
                opt_level=3):
            mod = seq(mod)
```

elvin-n commented on docs/how_to/deploy/adreno.rst:
```diff
@@ -0,0 +1,311 @@
+         mod = IRModule.from_expr(mod)
+         seq = tvm.transform.Sequential(
+            [
+                  relay.transform.InferType(),
```
Did you miss call of `ToMixedPrecision` transformation?

mbaret committed:
[AOT] Add CreateExecutorMetadata analysis pass (#13250)

AOT requires the ExecutorCodegenMetadata object to be
populated containing various pieces of information about
the compiled module. This commit adds a separate analysis
pass to create the metadata + some tests for the new pass.

In order to collect the device information correctly,
AOTLowerMain is extended to attach the device info as a
function attribute.

mehrdadh committed:
[microTVM][CRT][DOCS] Add a PyTorch tutorial for microTVM with CRT (#13324)

This commit adds a tutorial to compile and run a PyTorch model using   
microTVM, the AOT host-driven executor, and C runtime (CRT).

alanmacd committed:
[ci] Update Jenkins readme to match new directory structure (#13333)

Update Jenkins readme to match new directory structure

masahi committed:
[MetaSchedule] Fix the order of applying `AutoInline` in `ScheduleUsingAnchorTrace` (#13329)

* index on concat-fusion-fix: 3ffe5b12b fix te extern create_prim_func test

* Apply AutoInline to the last block after all other blocks are processed

* Do not require CanReverseComputeInline to be true when
CanComputeInline is false

* add comment

* add test

* cpplint

zxybazh committed:
[MetaSchedule] Add JSON Database Validation Scripts (#12948)

* Add validation scripts.

* Fix testing script.

* Fix lint.

* Fix lint.

* Fix inputs.

* Fix lint.

* Fix lint.

* Add timer func.

* Fix ci.

* Address comments.

* Add total time statistics.

* Fix lint.

vvchernov committed:
[QNN, ONNX] Extension of QLinearMatMul in ONNX front-end for all ranks of input tensors (#13322)

* QLinearMatMul was extended for all ranks of a and b

* CI test for QLinearMatMul was implemented (onnx front-end)

* fix after black check

* numpy type fix

* fix weight scale and zero point, output type

* fix after pylint

* resolve different input types in tests

* skip resolved TODO

* update covering of QLinearMatMul by tests

* pylint fixes

* skip test of QLinearMatMul on CUDA

Co-authored-by: Valery Chernov <valery.chernov@deelvin.com>

masahi committed:
[TIR] Check producer predicate in `ReverseComputeInline` (#13338)

* [TIR] Disallow reverse inline into a producer with non-trivial predicate

* add test

* Allow cases where the producer predicate can be implied by the new
predicate of the inlined block

* remove unused variable

* update comment in test to reflect the change in ReverseComputeInline

masahi committed:
[TOPI] Fix conv2d transpose for small channel (#13341)

* [TOPI] Fix conv2d transpose for small channel

* black

junrushao committed:
[Minor][Testing] Consolidate IRs into corresponding functions (#13339)

We moved most of the IR definition into the testing methods correspondingly.

Co-authored-by: Yaxing Cai <caiyaxing666@gmail.com>

srkreddy1238 committed:
[CPP_RPC][ANDROID] Fix cpp_rpc build failure (#13305)

* cpp_rpc build failure for Android devices with NDK version < 23

* * Make environment variable ANDROID_NDK_MAJOR optional.

Co-authored-by: Siva Rama Krishna Reddy B <sivb@blr-ubuntu-ripper.qualcomm.com>

csullivan committed:
[Hexagon] Make allocate_hexagon_array a hexagon contrib API (#13336)

Make 'allocate_hexagon_array' a hexagon contrib API

sergio-grovety committed:
[microNPU] Fixed MergeConstants pass on striped networks (#13281)

This PR fixes the bug in MergeConstants pass on striped networks on Ethos-U NPU.

The issue was caused by _DivideConstants_ pass which is introducing new mod parameters and changing their order. So ethosu_write parameter in some cases is moved from the end of the list to the middle.
E.g. from:
`[ethos-u_0_i0, p1, p2, p3, p4, p5, p6, ethosu_write]`
To:
`[ethos-u_0_i0, p1, p2, ethosu_write, placeholder, placeholder, placeholder, placeholder, placeholder, placeholder, placeholder, placeholder]`

Updated version of the  _GetArgsToMergeWithoutArgsNotInConstDict_ and _MakeNewConstDict_ methods in passes.cc can now correctly modify const_dict according to the new parameter list.

lhutton1 committed:
[TVMC] Global pass context for compile and tune (#13309)

* [TVMC] Global pass context for compile and tune

Comes as a followup from conversations in #13216. By making the pass
context a global value for both `compile` and `tune` commands, we can
ensure the pass context is exactly as the user expected and also
test components such as `convert_graph_layout` under a pass context
suitable for testing (e.g. add instruments). With this change, it
becomes the users responsibility to ensure the PassContext they
select is suitable for the passes that will be run. By default,
`opt_level` remains as 3 so current workflows that do not alter the pass
context from the command line / TVMC Python API should not be affected.

Change-Id: I7a601daf6fbe664f77bce1b45efeb7ca29f621b3

* fix vitis-ai test and typo

Change-Id: I04f5bd031ae4717825f42e373bcb0e1e2c1c9d90

nverke committed:
[TIR] Update ReductionIterNotIndexOutputBuffer to check BlockRealizeN… (#13301)

* [TIR] Update ReductionIterNotIndexOutputBuffer to check BlockRealizeNodes match_buffer statements when validating writes

* Add test to verify that tensorized blocks are properly validated

* update to take into account all match buffer regions.

* lint

mehrdadh committed:
[Docker]Refactor timezone script and NRF installation (#13342)

This PR refactors timezone setup to a separate script that docker/install/ubuntu_install_core.sh
Also, it adds a script to install NRF and reused in both cortexm docker and RVM installation path.

wzh99 committed:
[TIR][Arith] Fix divisor checking in `TryConstFold` (#13348)

Fix denominator checking in `TryConstFold`.

zxybazh committed:
[MetaSchedule][Minor] Fix Typo in ApplyCustomRule Schedule Rule (#13353)

* Fix typo.

* Add regression test.

masahi committed:
[MetaSchedule] Improve inlining and `VerifyGPUCode` for quantized model workload (#13334)

* [MetaSchedule] Add a new schedule rule to inline all scalar constants

* add doc

* reorg

* identify constant block by its structure, not by name

zxybazh committed:
[MetaSchedule][Minor] Allow Zero Run Time In Benchmarking Result (#13354)

This PR introduces a check to prevent records with run time of zero into the training data of cost model. This is because when working on microTVM there're cases where the run time of certain successful runs is very tiny, such that it got recorded as zero. In such cases, the runtime of 0 would break XGBoost model because it introduces infinite running speed in GFLOPs. A regression test was also added.

sunggg committed:
[Bugfix][TIR] Patch for PR#13269 to support Python 3.10 (#13350)

It seems like there is some inconsistency across the python versions and make PR https://github.com/apache/tvm/pull/13269 fails at Python 3.10. 
This patch fixes this issue. 

Co-authored-by: Junru Shao <junrushao1994@gmail.com>

masahi committed:
[MetaSchedule] Unannotate `schedule_rule` if corresponding schedule func is not found (#13346)

yelite committed:
[MetaSchedule] Skip empty fx graph in TorchBench tuning script (#13356)

masahi committed:
[MetaSchedule] Fuse loops around shared to global store block in `MultiLevelTilingTensorCore` (#13357)

* Fuse shared to global store loops in MultiLevelTilingTensorCore

* update test

wrongtest-intellif committed:
[TIR][Schedule] Make consistent implementation for GetProducers() & GetConsumers() (#13344)

Currently there are two versions of `GetConsumers()` and `GetProducers()` implementation. Make them consistent to avoid possible bug when there are WAR dependencies.

StrongerXi committed:
[Relay] Refactor constant folding over expr into a utility function (#13343)

junrushao committed:
[TIR] Make syntax of AST nodes different than ops (#13358)

As part of effort of more formal TIR semantics, we want to more
explicitly differentiate TIR AST nodes (defined in `tir/expr.h`)
and TIR ops (defined in `tir/op.h`).

A naming convention is that:
- Lowercased methods, for example, `tvm.tir.mul`, means an TIR op, which
  will be eagerly constant-folded, i.e. `mul(1, 2)` returns `3`
  immediately rather than creating an AST node.
- Capitalized callable, for example, `Mul`, means creating an AST node
  without constant folding.

This PR makes this behavior more explictly by printing `T.Mul(a, b)`
directly when `a` and `b` are both constants, rather than sugaring it
into `mul(a. b)` or `a * b`, so that the difference between an op and
an AST node is clarified.

Co-authored-by: Yaxing Cai <caiyaxing666@gmail.com>

Co-authored-by: Yaxing Cai <caiyaxing666@gmail.com>

ibsidorenko committed:
[FQ2I] Add cast back to input data type after AvgPool2d (#13332)

[FQ2I] Add cast back to output data type after AvgPool2d

This commit fixes the following issue:
For the sequence of qnn.dequantize -> avg_pool2d -> conv2d ->
qnn.quantize FQ2I pass inserts qnn.requantize (or cast) to int32
unconditionally before AvgPool2d. As a result fake quantized qnn.conv2d
gets input as int32 dtype, but it is forbidden for qnn.conv2d (supports
only uint8/int8/int16).

This commit adds the following:
Add cast back to output data type after AvgPool2d. This preserve input
dtype == output dtype for this op.

dsbarinov1 committed:
ToMixedPrecision fix

dsbarinov1 commented:
@elvin-n please, review final adjustments :)

junrushao committed:
[IRBuilder][Minor] Add intrinsics like `T.int32x4` (#13361)

This PR adds all common TIR intrinsics like `T.int32x4`, `T.floatx4`.

Co-authored-by: Yaxing Cai <caiyaxing666@gmail.com>

multiverstack-intellif committed:
[TIR][Schedule] Fix cache_read loc detecting and region_cover checking (#13345)

Fix 2 issues of cache related primitives:
*  Fix region_cover checking for cache related primitives
*  Fix CacheLocDetector for nested SeqStmt

Co-authored-by: Min Chen <chen.min@intellif.com>

junrushao committed:
[TVMScript] Reorganize the folder structure (#12496)

This PR introduces some minor restructuring of the `python/tvm/script`
folder structure to make it more convenient for future upstreaming.

Co-authored-by: Yaxing Cai <caiyaxing666@gmail.com>

gigiblender committed:
[ci] Assert some tests are not skipped in the CI (#12915)

In this PR, the skipped tests script will also check if tests in the `required_tests_to_run.json` have not been skipped. If there are skipped tests, they will be added to the returned comment. 

I am not entirely sure where it's best to place the `required_tests_to_run` file, so I left it in `tvm/ci/scripts/`. I am happy to take suggestions.

Aims to prevent situations such as #12529

gigiblender committed:
[CI] Separate the ci scripts into Github and Jenkins scripts (#13368)

This PR is a duplicate of #12940 and #12941. For some reason, I am unable to reopen #12940.

Lunderberg committed:
[TIR][Bugfix] Fix AXIS_SEPARATORS in tir.Schedule.transform_layout (#13326)

Preivously, the block SREF reuse only included a single step of
changes, and would have an incorrect mapping if multiple sequential
changes to the TIR block occurred.  This could happen if a
`BufferStore` was updated, followed by replacement of `Block` iter
vars/values.  This commit tracks the Block replacements across each
usage, to ensure the SREF instances remain valid.

driazati committed:
[ci] Fix Jenkins quoting (#13380)

Merging #13368 caused CI to pass but run more than it needed to due to
some failures in determination. This fixes the interpolation to use `"`
which should correctly pass through the variables

Co-authored-by: driazati <driazati@users.noreply.github.com>

gigiblender committed:
[CI] Do not merge before running CI on main (#13372)

This PR does not merge `main` if CI is running already on `main`. It aims to avoid a case where a race happens between two subsequent commits, and one of them merges the other.

Fixes #12392.

vinx13 committed:
[Codegen] Fix CUDA codegen for int64 Ramp (#13382)

zxybazh committed:
[MetaSchedule] Add `from-target` Defaults for LLVM VNNI Targets (#13383)

leandron committed:
[TFLite] Enable int64 biases for int16 quantized operators (#12042)

This enables int64 biases for quantized fully connected, requantize
and transpose convolution in TFLite networks. It goes on top of existing
int16 support for TFLite frontend.

Add a test case using DS_CNN int16 quantized.
